enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,328 @@
+  (BMO). Uses the [Ironic agent][ipa] (also known as ironic-python-agent or
+  simply IPA) for writing the CoreOS image.
+
+- ZTP (Zero-Touch Provisioning) flow using the so-called *live ISO* Ironic and
+  baremetal-operator feature, the assisted service and the assisted agent.
+
+## Motivation
+
+Having two different provisioning flows comes with significant maintenance and
+development costs:
+
+- Some features, namely firmware and RAID configuration, require using the
+  Ironic agent, which is not present in the *live ISO* flow.
+
+- Supporting (i)PXE boot with the *live ISO* flow requires creating new API
+  in baremetal-operator specifically for this purpose.
+
+- In the *live ISO* flow, it is not possible to notify Ironic (and thus
+  baremetal-operator) when the actual installation finishes. From the
+  perspective of Metal3 components, the installation is finished once the ISO
+  is connected. As a result, Ironic cannot disconnect the ISO from the
+  machine's BMC, nor can it update the BMC's boot device settings properly.
+
+- The *live ISO* flow disables Ironic *inspection* process and has to simulate
+  it by updating the `BareMetalHost` based on the assisted service's inspection
+  process. This creates an unnessesary coupling between the assisted service
+  and the baremetal-operator's hardware inventory format (which, in turn, is
+  derived from the Ironic's inventory format).
+
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+
+- Any different path through the `BareMetalHost` state machine makes it
+  exponentially more difficult for upstream developers to reason about the
+  effects of any changes. In addition, the *live iso* flow is not tested
+  upstream by the Metal3 project.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+- Continue to allow non-ZTP assisted deployments without depending on Metal3.
+- Not depend on assisted components in the core metal platform
+
+### Non-Goals
+
+- Converging two agent implementations (Ironic and assisted) into a single one.
+- Supporting per-host network configuration through the BMH
+  PreprovisioningNetworkData API in the ZTP case.
+- Performing firmware config changes only after the assisted service runs host
+  validations.
+
+## Proposal
+
+- Building CoreOS images for the *standard* flow is handled by
+  [image-customization-controller][icc] (ICC). ICC will start ignoring all
+  `PreprovisioningImage`s that have the `infraenvs.agent-install.openshift.io`
+  label (copied from the corresponding `BareMetalHost` by BMO).
+
+- A new controller will be created as part of the assisted service to
```

Comment:
This might be out of scope for this enhancement, but I'd consider putting it in the assisted-image-service (rather than assisted-service) so it has direct access to the image-building APIs. This could potentially avoid e.g. the need to pass the ironic agent config in the InfraEnv spec that is supposed to belong to the user.

Reply:
Good point, we should consider that.  Also agree that it probably doesn't need to be included here though.

Reply:
The ignition config for the assisted discovery iso is defined in the assisted-service.
We can definitely add the ironic agent config to it.
Unlike our initial thought there is no need to add the ironic config to the infraEnv